### PR TITLE
Introduce static elementary types to use

### PR DESCRIPTION
**Pull request by lastmjs**
### Checklist
- [x] Code compiles correctly
- [ ] All tests are passing (except linter, still need help on that)
- [x] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary (I don't believe there is anything to change here)
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description

Completing https://github.com/ethereum/solidity/issues/4502

### CODE REVIEW

**Comment by lastmjs** [comment]
I'm kind of confused by the code coverage failing here, can someone point me in the right direction?

**Comment by chriseth** [comment]
Thanks for your contribution! While reviewing, @ekpyron noted that the static objects have to be references to shared pointers, because otherwise `shared_from_this` would not work. In general, we should strive to disallow bare objects of types at all, which is of course out of scope for this PR.

One way to both correct the problem that these objects here have to be shared pointers and go into the direction of disallowing bare objects would be the following:

Instead of a struct containing these static members, add public static functions to the types which have a static local variable of shared pointer type that return a reference to the object.

So it would e.g. be:

```
class IntegerType
{
...
static IntegerType& address() { static shared_ptr<IntegerType> addr(new IntegerType(160, Modifier::Address)); return *addr; }
}
```

Using functions also has the advantage that we get around the static initialization order problem.

**Comment by lastmjs** [comment]
I have done your bidding. Let me know what you think.

**Comment by ekpyron** [comment]
This looks good, but unfortunately, while you were working on this PR, #4887 was merged, which is now in conflict with this PR - do you think you could rebase this PR and adjust it to these latest changes in ``develop``? Instead of ``IntegerType::address()`` it will now have to be ``AddressType::address()``.

**Comment by ekpyron** [comment]
Rebasing will probably also fix the external test runs, which break due to breaking changes in separately merged PRs.

**Comment by lastmjs** [comment]
I'm on it

**Comment by lastmjs** [comment]
Ready for final review

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/ast/Types.h [hunk 1]
```diff
@@ -319,6 +319,8 @@ class Type: private boost::noncopyable, public std::enable_shared_from_this<Type
 class AddressType: public Type
 {
 public:
+    static AddressType& address() { static std::shared_ptr<AddressType> addr(new AddressType()); return *addr; }
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: Uses spaces for indentation.

Comment (chriseth) [reply libsolidity/ast/Types.h]: Also please don't use `new`. Use `make_shared` instead.

**Review by chriseth — CHANGES_REQUESTED** [review]
Please do the small changes and then squash everything into a single commit. The first commit still contains merge conflict markers.

**Comment by ekpyron** [comment]
Unfortunately, this will need to be rebased again, if #4911 is merged before it.

File: libsolidity/ast/Types.h (hunk 1 above)
Comment (lastmjs) [reply libsolidity/ast/Types.h]: Am I not using spaces for indentation? My editor is set to use 4 spaces for tabs, I'm trying to change that line but I think it's already spaces

Comment (chriseth) [reply libsolidity/ast/Types.h]: I'm sorry, this was a little unclear: You should use tabs everywhere, but these are spaces.

Comment (lastmjs) [reply libsolidity/ast/Types.h]: Got it

**Comment by lastmjs** [comment]
It looks like the `AddressType` constructor has a new `StateMutability` parameter. What should I pass in here when creating the shared pointer?

```c++
static AddressType& address() { static std::shared_ptr<AddressType> addr(std::make_shared<AddressType>()); return *addr; }
```

It seems like the `address` function should take the `StateMutability` as a parameter. In that case, what do I pass in when calling `address`?

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -2131,7 +2131,7 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 				"after argument-dependent lookup in " + exprType->toString() +
 				(memberName == "value" ? " - did you forget the \"payable\" modifier?" : ".");
 		if (exprType->category() == Type::Category::Contract)
-			for (auto const& addressMember: AddressType(StateMutability::Payable).nativeMembers(nullptr))
+			for (auto const& addressMember: AddressType::address(StateMutability::Payable).nativeMembers(nullptr))
```
Comment (lastmjs) [inline libsolidity/analysis/TypeChecker.cpp]: I have no idea what the `StateMutability` parameter should be here

File: libsolidity/ast/Types.h [hunk 3]
```diff
@@ -1,18 +1,18 @@
 /*
-    This file is part of solidity.
+	This file is part of solidity.
```
Comment (lastmjs) [inline libsolidity/ast/Types.h]: I just blanket-converted all spaces to tabs in this file, did we also want the comment to use tabs?

Comment (chriseth) [reply libsolidity/ast/Types.h]: This is a very old file, please just keep existing lines in comments.

**Comment by chriseth** [comment]
Please create two functions: One just called `address` with state mutability "nonpayable" and one called `addressPayable` with state mutability 'payable'.

**Comment by lastmjs** [comment]
When should I use each? There are a couple of instances in this pull request where I need to decide

**Comment by chriseth** [comment]
The current code should already have them.

**Comment by chriseth** [comment]
Can you please squash the commits into one?

**Review by ekpyron — APPROVED** [review]

**Comment by lastmjs** [comment]
I'm not sure what the test failures are, tests look fine on my machine...oh wait, looks like some package.json versioning issue

**Review by chriseth — APPROVED** [review]

**Pull request merged**